### PR TITLE
Create service order

### PR DESCRIPTION
**Pull request by Manuv46**

### CODE REVIEW

File: src/functions/deleteOrder/handler.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * In this case event.body type is type of 'Order'
+*/
+const deleteOrderHandler: ValidatedEventAPIGatewayProxyEvent<void> = async (event) => {
+    let res: Response<Order>;
+    try {
+
+        const user: AuthenticatedUser = await AuthenticatedUser.fromToken(event.headers?.AccessToken);
+
+        res = Response.fromData<Order>(
+            await deleteOrder(event.pathParameters.id),
```
Comment (URBA22) [inline src/functions/deleteOrder/handler.ts]: This is very unsafe. 
This function allow every authenticated user to delete every order, with no check on userId.
It also allow the user to delete a 'completed' order, but a user can delete the order only if the order is not completed yet. If we don't check the status, we're going to lose the story of user's orders.
Important: a vendor can never be able to delete or modify an order, or he could scam customers! (For example he change the price in the order).
Same whit the update function.

File: src/functions/getOrder/handler.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/*
+ * Remember: event.body type is the type of the instantiation of ValidatedEventAPIGatewayProxyEvent
+ * In this case event.body type is 'void' -> we have no body!
+*/
+const getOrderHandler: ValidatedEventAPIGatewayProxyEvent<void> = async (event) => {
+    let response: Response<Order>;
+    try {
+        const user: AuthenticatedUser = await AuthenticatedUser.fromToken(event.headers?.AccessToken);
+        response = Response.fromData<Order>(
+            await getOrder(event.pathParameters?.id),
```
Comment (URBA22) [inline src/functions/getOrder/handler.ts]: You have to check if the order userId is the same as `await user.getUserId()` or a user could see a different user order (if he is not a vendor).

File: src/functions/getOrder/function.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import dbContext from '@dbModel/dbContext';
+import Order from '@dbModel/tables/order';
+
+const getOrder = async (id: string): Promise<Order> => {
```
Comment (URBA22) [inline src/functions/getOrder/function.ts]:
```suggestion
const getOrder = async (id: string, userId: string): Promise<Order> => {
```

File: src/functions/scanOrder/function.ts [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
-import Product from '@dbModel/tables/product';
+import Order from '@dbModel/tables/order';
 import { ScanOptions } from '@aws/dynamodb-data-mapper';
 import { objectToConditionExpression } from 'utilities-techsweave';
 
 
-const scanProduct = async (filter: any): Promise<{
-    items: Product[],
-    lastKey: Partial<Product>
+const scanOrder = async (filter: any): Promise<{
```
Comment (URBA22) [inline src/functions/scanOrder/function.ts]:
```suggestion
const scanOrder = async (filter: any, userId?: string): Promise<{
```

File: src/functions/scanOrder/function.ts [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
+    items: Order[],
+    lastKey: Partial<Order>
 }> => {
-    let items: Product[] = [];
-    let lastKey: Partial<Product>;
+    let items: Order[] = [];
+    let lastKey: Partial<Order>;
```
Comment (URBA22) [inline src/functions/scanOrder/function.ts:12]:
```suggestion
    let lastKey: Partial<Order>;
    let filter = await objectToConditionExpression(filter.filter);
    if (userId) {
        filter = {
            type = 'And',
            conditions: [{
                    type = 'Equals',
                    subject = 'userId',
                    object = userId
                },
                filter 
            ]
    }
```
Maybe you have to switch subject and object, I don't remebember.

File: src/functions/scanOrder/function.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ const scanProduct = async (filter: any): Promise<{
         filter: await objectToConditionExpression(filter.filter)
```
Comment (URBA22) [inline src/functions/scanOrder/function.ts]:
```suggestion
        filter: filter
```

File: src/functions/scanOrder/handler.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+/*
+ * Remember: event.body type is the type of the instantiation of ValidatedEventAPIGatewayProxyEvent
+ * In this case event.body type is type of 'Order'
+*/
+const scanOrderHandler: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event) => {
+    let res: Response<Order> = new Response<Order>();
+
+    try {
+        //const user: AuthenticatedUser = await AuthenticatedUser.fromToken(event.headers?.AccessToken);
```
Comment (URBA22) [inline src/functions/scanOrder/handler.ts]:
```suggestion
        const user: AuthenticatedUser = await AuthenticatedUser.fromToken(event.headers?.AccessToken);
```

File: src/functions/scanOrder/handler.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
```
Comment (URBA22) [inline src/functions/scanOrder/handler.ts]:
```suggestion
let userId = null;
if (!(await user.isVendor()))
    userId = await userId.getUserId();
```

File: src/functions/scanOrder/handler.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+        const result = await scanOrder(event.body);
```
Comment (URBA22) [inline src/functions/scanOrder/handler.ts]:
```suggestion
        const result = await scanOrder(event.body, userId);
```
Same logic of get:
- If user is customer: he can see only orders that belong to him
- If user is vendor: he can see all orders

I suggest some ideas about how implement this logic.

**Review by URBA22 — CHANGES_REQUESTED** [review]

**Review by URBA22 — APPROVED** [review]

**Review by samupd2013 — APPROVED** [review]

**Pull request merged**